### PR TITLE
Update to use GraphStage

### PR DESCRIPTION
**Pull request by jasonmartens**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,8 +1,8 @@
-name := "AkkaStreamsCommandPipe"
+name := "shell-command-flow"
```
Comment (jiahuang) [inline build.sbt:1]: can you change the repo name to reflect the package name change?

File: build.sbt [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 
 organization := "io.dronekit"
 
-version := "1.0"
+version := "1.1-M2"
```
Comment (jiahuang) [inline build.sbt]: Does M2 need to be in the package version? Why not just 1.0 since this is a different package now?

Comment (jasonmartens) [inline build.sbt]: Trying to keep track of compatibility changes for the streams library. Will remove the M when we can depend on streams 2.0 release.

File: build.sbt (hunk 1 above)
Comment (jasonmartens) [inline build.sbt:1]: Yeah, ok.

File: src/main/scala/Main.scala [hunk 3]
```diff
@@ -2,7 +2,9 @@ import akka.actor.ActorSystem
 import akka.stream.ActorMaterializer
 import akka.stream.scaladsl.{Sink, Source}
 import akka.util.ByteString
-import io.dronekit.cloud.ExternalCommandFlow
+import com.typesafe.scalalogging.Logger
```
Comment (jiahuang) [inline src/main/scala/Main.scala:5]: Do we need a Main? This seems like it's an integration test.

Comment (jasonmartens) [inline src/main/scala/Main.scala:5]: Dunno, it's kind of handy to have a complete example of how to use it. But yes, maybe it should be moved to an IT folder.

File: src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            outputClosed.set(true)
+            pushAndComplete()
+
+        }
+
+        setHandler(in, new InHandler{
+          override def onPush: Unit = {
+            val elem = grab(in)
+            if (!errorQueue.isEmpty) {
+              // What is the right thing to do if there is stderr output?
```
Comment (jiahuang) [inline src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala:107]: This is out of scope for this PR but eventually the user should be able to consume stderr as a different stream and decide what to do with it.

File: src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+          while (!inputQueue.isEmpty && !errorAbort.get) {
+            in.write(inputQueue.takeLast().toArray[Byte])
+          }
+          try {
+            in.close()
+          } catch {
+            case ex: IOException if ex.getMessage == "Stream closed" =>
+              log.error(s"Ignoring $ex in processInput while closing")
+            case ex: IOException if ex.getMessage == "Broken pipe" =>
+              log.error(s"Ignoring $ex in processInput while closing")
```
Comment (jiahuang) [inline src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala:159]: There's two case statements that do the same thing

Comment (jasonmartens) [inline src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala:159]: Yes, but they are handling two different cases.

File: src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala (hunk 4 above)
Comment (jasonmartens) [inline src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala:107]: agree, I'll make an issue.

File: src/test/scala/ShellCommandFlowSpec.scala [hunk 6]
```diff
@@ -15,23 +15,23 @@ import scala.language.postfixOps
...
     val result = Await.result(stream, 1 seconds)
-    result.head shouldBe (1 to 10).mkString(" ") + " "
+    result.head shouldBe (1 to 100).mkString(" ") + " "
   }
 
   it should "throw a NoSuchElementException if the source is empty" in {
     val source = Source.empty[ByteString]
-    val stream = source.via(ExternalCommandFlow(Seq("/bin/cat"))).grouped(1000000).runWith(Sink.head)
+    val stream = source.via(ShellCommandFlow(Seq("/bin/cat"))).grouped(1000000).runWith(Sink.head)
     intercept[NoSuchElementException] {
```
Comment (jiahuang) [inline src/test/scala/ShellCommandFlowSpec.scala:32]: does this test fail if the `NoSuchElementException` is never thrown?

Comment (jasonmartens) [inline src/test/scala/ShellCommandFlowSpec.scala:32]: Yes

File: src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala (hunk 5 above)
Comment (jiahuang) [inline src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala:159]: does this work? `case ex: IOException if (ex.getMessage == "Broken pipe" || ex.getMessage == "Stream closed")`

Comment (jasonmartens) [inline src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala:159]: Probably? Is that much better?

**Pull request merged**